### PR TITLE
single-node-multi-gpu Pagerank beta

### PR DESCRIPTION
**Pull request by afender**

### CODE REVIEW

File: conda/environments/cugraph_dev.yml [hunk 1]
```diff
@@ -1,12 +1,16 @@
 name: cugraph_dev
 channels:
 - nvidia
-- rapidsai
+- rapidsai-nightly
```
[the remark this answers is not in the document]
Comment (afender) [reply conda/environments/cugraph_dev.yml]: It good to know that we have both env now in cugraph. I'll add that in this branch as well.

File: cpp/src/multi_pagerank/cuda_kernels.cu [hunk 2]
```diff
@@ -0,0 +1,2350 @@
+/*
+ * Copyright (c) 2016, NVIDIA CORPORATION. All rights reserved.
```
[the remark this answers is not in the document]
Comment (afender) [reply cpp/src/multi_pagerank/cuda_kernels.cu]: Added to the list of known issues for 0.7

File: cpp/src/multi_pagerank/cuda_kernels.cu [hunk 3]
```diff
@@ -0,0 +1,2350 @@
...
+	data.w = (data.w - 0x01010101) & ~data.w & 0x80808080;
+	data.w = __ffs(data.w);
+	offs.w = data.w ? tid*sizeof(data) + 12 + ((data.w-1)>>3) : -1;
+	bl_off[tid] = offs;
+	return;
+}
+*/
+
+template<typename INT_T>
+__device__ const char *str2bin(const char *s, INT_T *v) {
```
[the remark this answers is not in the document]
Comment (afender) [reply cpp/src/multi_pagerank/cuda_kernels.cu]: Added to the list of future investigations

[the remark this answers is not in the document]
Comment (afender) [reply cpp/src/multi_pagerank/cuda_kernels.cu]: This doesn't seem to create problems but no one is familiar with this function so we should double check it.

File: cpp/src/multi_pagerank/cuda_kernels.cu [hunk 4]
```diff
@@ -0,0 +1,2350 @@
...
+		cubSortKeys(cols_d[nextp], cols_sort_d, nnz[nextp]);
+		cubUnique(cols_sort_d, recvRows_m, n_d, nnz[nextp]);
+		CHECK_CUDA(cudaMemcpy(rcount + nrecv,
+				      n_d,
+				      sizeof(*rcount),
+				      cudaMemcpyDeviceToHost));
+
+		MPI_Sendrecv(rcount + nrecv, 1, MPI_LONG_LONG, nextp, TAG(rank),
+			     scount + nsend, 1, MPI_LONG_LONG, prevp, TAG(prevp),
+            		     COMM, MPI_STATUS_IGNORE);
```
[the remark this answers is not in the document]
Comment (afender) [reply cpp/src/multi_pagerank/cuda_kernels.cu]: MPI errors should be fatal by default. 
There is a way for MPI to return an error and for the program to handle them. This is a better design for a library. I added it to the list of future improvements.

File: conda/environments/cugraph_dev.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (afender) [reply conda/environments/cugraph_dev.yml]: Addressed in 7bcda01bf704da4140704ffd628cc77c43589aa1

File: python/cugraph/mg_pagerank/mg_pagerank_wrapper.pyx [hunk 5]
```diff
@@ -0,0 +1,40 @@
+from libc.stdint cimport uintptr_t
+from c_mg_pagerank cimport *
+
+def mg_pagerank(input_df, pr_col_length, global_v):
+    source_col = input_df[input_df.columns[0]]
+    dest_col = input_df[input_df.columns[1]]
+    #pagerank_col = cudf.Series(np.ones(pr_col_length, dtype=np.float32))
+    #vertices_col = cudf.Series(np.ones(pr_col_length, dtype=np.int32))
+    cdef gdf_column* vid_ptr= <gdf_column*>malloc(sizeof(gdf_column))
+    cdef gdf_column* pr_ptr= <gdf_column*>malloc(sizeof(gdf_column))
```
[the remark this answers is not in the document]
Comment (afender) [reply python/cugraph/mg_pagerank/mg_pagerank_wrapper.pyx]: @Iroy30

File: dask/test_dask_cugraph_sort_values_binned.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+from dask.distributed import Client
+
+import dask_cudf
+
+
+# 2. Set the Number of GPU Devices and File Paths
+
+number_of_devices = 2
+scheduler_file_path = r"/home/seunghwak/cluster.json"
+input_data_path = r"/home/seunghwak/TMP/Input-small/edges"
```
[the remark this answers is not in the document]
Comment (afender) [reply dask/test_dask_cugraph_sort_values_binned.py]: @seunghwak

File: dask/test_dask_cugraph_sort_values_binned.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    all_devices = list(range(n))
+    visible_devices = ",".join(map(str, all_devices[i:] + all_devices[:i]))
+    os.environ["CUDA_VISIBLE_DEVICES"] = visible_devices
+
+
+# 4. Create a Client
+
+print("Initializing.")
+
+client = Client(scheduler_file="/home/seunghwak/cluster.json",
```
[the remark this answers is not in the document]
Comment (afender) [reply dask/test_dask_cugraph_sort_values_binned.py]: @seunghwak

File: dask/test_dask_cugraph_pagerank.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    os.environ["CUDA_VISIBLE_DEVICES"] = visible_devices
+
+
+# 4. Create a Client
+
+print("Initializing.")
+
+start_time = time.time()  # start timing from here
+
+client = Client(scheduler_file="cluster.json",
```
[the remark this answers is not in the document]
Comment (afender) [reply dask/test_dask_cugraph_pagerank.py]: @Iroy30

**Comment by afender** [comment]
@aschaffer I completely agree. The style is more C than C++. A good code clean-up / refactoring should take advantage of C++ features to make the code more robust and maintainable. I think we need weeks to do that properly. I also added a task to search and remove all dead code.

File: cpp/src/multi_pagerank/cuda_kernels.cu (hunk 3 above)
[the remark this answers is not in the document]
Comment (afender) [reply cpp/src/multi_pagerank/cuda_kernels.cu]: This seems to be dead kernel actually as @kaatish added a CPU reader based on STL.

File: python/cugraph/mg_pagerank/mg_pagerank_wrapper.pyx (hunk 5 above)
[the remark this answers is not in the document]
Comment (Iroy30) [reply python/cugraph/mg_pagerank/mg_pagerank_wrapper.pyx]: Added freeing of the gdf_columns `a11cf22`

File: dask/test_dask_cugraph_pagerank.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Iroy30) [reply dask/test_dask_cugraph_pagerank.py]: changed by @seunghwak

**Pull request merged**